### PR TITLE
MM-T3187 Unread channels sort at top

### PR DESCRIPTION
**Pull request by clarmso**
#### Summary

Automate the test MM-T3187.

#### Ticket Link

https://automation-test-cases.vercel.app/test/MM-T3187

#### Checklist

- [x] Added or updated unit tests (required for all new features)
- [ ] Has UI changes
- [ ] Includes text changes and localization file updates

#### Device Information
This PR was tested on:  

Android 10.0+ (R)

#### Screenshots

N/A

### CODE REVIEW

File: detox/e2e/test/messaging/message_posting.e2e.js [hunk 1]
```diff
@@ -26,4 +29,23 @@ describe('Messaging', () => {
 
         await expect(element(by.text(text))).toExist();
     });
+
+    it('MM-T3187 Unread channels sort at top', async () => {
+        // # Open channel drawer (with at least one unread channel)
+        await element(by.id('channel_drawer_button')).tap();
+
+        // # Verify unread channel(s) display at top of channel list (with mentions first, if any), in alphabetical order, with title "Unreads"
+        await expect(element(by.text('UNREADS'))).toBeVisible();
```
Comment (clarmso) [inline detox/e2e/test/messaging/message_posting.e2e.js:38]: I'm not sure how to ensure the channels are in alphabetical order.

File: detox/e2e/test/messaging/message_posting.e2e.js [hunk 2]
```diff
@@ -26,4 +29,23 @@ describe('Messaging', () => {
...
+        // # Tap an unread channel to view it
+        await element(by.text(newChannel.display_name)).tap();
+
+        // # Open channel drawer again
+        await element(by.id('channel_drawer_button')).tap();
+
+        // * Channel you just read is no longer listed in Unreads
+        await expect(element(by.text('PUBLIC CHANNELS'))).toBeVisible();
+        await expect(element(by.text('UNREADS'))).not.toBeVisible();
+        await expect(element(by.text(newChannel.display_name)).atIndex(1)).toBeVisible();
```
Comment (clarmso) [inline detox/e2e/test/messaging/message_posting.e2e.js:49]: There's `newChannel.display_name` on the navbar and the channel list. I've tried `withAncestor(by.type('SectionList'))` but no avail.  Using `atIndex(1)` is not good. If there's an XPath-like syntax to specify the channel name from the channe list, please let me know.

**Comment by enahum** [comment]
Snapshot tests are failing

```
Summary of all failing tests
 FAIL  app/screens/channel/channel_nav_bar/channel_drawer_button/channel_drawer_button.test.js
  ● ChannelDrawerButton › should match, full snapshot

    expect(received).toMatchSnapshot()

    Snapshot name: `ChannelDrawerButton should match, full snapshot 1`

    - Snapshot  - 0
    + Received  + 1

    @@ -20,10 +20,11 @@
                "justifyContent": "center",
                "paddingHorizontal": 10,
              },
            ]
          }
    +     testID="channel_drawer_button"
        >
          <View>
            <View
              style={
                Object {

      54 | 
      55 |         // no badge to show
    > 56 |         expect(wrapper.getElement()).toMatchSnapshot();
         |                                      ^
      57 |         expect(wrapper.find(Badge).length).toEqual(0);
      58 | 
      59 |         // badge should render

      at Object.<anonymous> (app/screens/channel/channel_nav_bar/channel_drawer_button/channel_drawer_button.test.js:56:38)

  ● ChannelDrawerButton › should match, full snapshot

    expect(received).toMatchSnapshot()

    Snapshot name: `ChannelDrawerButton should match, full snapshot 2`

    - Snapshot  - 0
    + Received  + 1

    @@ -20,10 +20,11 @@
                "justifyContent": "center",
                "paddingHorizontal": 10,
              },
            ]
          }
    +     testID="channel_drawer_button"
        >
          <View>
            <Icon
              allowFontScaling={false}
              color="#ffffff"

      59 |         // badge should render
      60 |         wrapper.setProps({badgeCount: 1, visible: true});
    > 61 |         expect(wrapper.getElement()).toMatchSnapshot();
         |                                      ^
      62 |         expect(wrapper.find(Badge).length).toEqual(1);
      63 |     });
      64 | 

      at Object.<anonymous> (app/screens/channel/channel_nav_bar/channel_drawer_button/channel_drawer_button.test.js:61:38)


Snapshot Summary
 › 2 snapshots failed from 1 test suite. Inspect your code changes or run `npm test -- -u` to update them.
```

File: detox/e2e/test/messaging/message_posting.e2e.js (hunk 2 above)
Comment (josephbaylon) [reply detox/e2e/test/messaging/message_posting.e2e.js:49]: Hi @clarmso, I think `by.type()` assumes the native class type from iOS or Android (https://github.com/wix/Detox/blob/master/docs/APIRef.Matchers.md#bytypeclassname). Maybe a workaround is to add `id` to `SectionList` and use `withAncestor(by.id(...))` instead. I haven't tried it yet but maybe worth to give a shot.

File: detox/e2e/test/messaging/message_posting.e2e.js (hunk 1 above)
Comment (josephbaylon) [reply detox/e2e/test/messaging/message_posting.e2e.js:38]: Hi @clarmso I tried adding `testID='channel_item_display_name'` to the ChannelItem > render > `Text` element of `app/components/sidebars/main/channel_list/channel_item/channel_item.js`

and then added this verification, which works,
```
await expect(element(by.id('channel_item_display_name')).atIndex(0)).toHaveText(newChannel.display_name);
```

You can probably add two channels, channel1 = 'B Channel', channel2 = 'A Channel', then have the following verification for order,

```
await expect(element(by.id('channel_item_display_name')).atIndex(0)).toHaveText(channel2.display_name);
await expect(element(by.id('channel_item_display_name')).atIndex(1)).toHaveText(channel1.display_name);
```

Comment (clarmso) [reply detox/e2e/test/messaging/message_posting.e2e.js:38]: If multiple elements on the same screen can have the same ID, I'm ok. (For web, it's not a good thing to have the same ID used in multiple elements.) 

I'll update the snapshot if necessary. :)

Comment (enahum) [reply detox/e2e/test/messaging/message_posting.e2e.js:38]: You can construct the id to be something like `channel_item.${channel.id}` to avoid repetition if the same id

File: detox/e2e/test/messaging/message_posting.e2e.js (hunk 2 above)
Comment (clarmso) [reply detox/e2e/test/messaging/message_posting.e2e.js:49]: Adding testID to `SectionList` does not work for me, but I could do so with `ChannelList`.  Still, I can't get `withDescendants` and `withAncestor` working. I've tried the following:

`await expect(element(by.id('channels_list').withDescendant(by.id('channel_item_display_name')))).toExist();`
`await expect(element(by.text(newChannel.display_name).withAncestor(by.id('channels_list')))).toExist();`

Turning on verbose (on --loglevel) does not show info.

In addition, I've tried the following. `element(by.id('channel_item_display_name')).atIndex(0)`) points to "Off-Topic" despite react devtools shows that index 0 points to "Channel <random>" 

`await expect(element(by.id('channel_item_display_name')).atIndex(0)).toHaveText(newChannel.display_name);`

File: detox/e2e/test/messaging/message_posting.e2e.js (hunk 1 above)
Comment (clarmso) [reply detox/e2e/test/messaging/message_posting.e2e.js:39]: @enahum Good point. If unique ID isn't required for mobile, I'd leave it as is for now.

File: detox/e2e/test/messaging/message_posting.e2e.js [hunk 3]
```diff
@@ -26,4 +29,26 @@ describe('Messaging', () => {
...
     });
+
+    it('MM-T3187 Unread channels sort at top', async () => {
+        // # Open channel drawer (with at least one unread channel)
+        await element(by.id('channel_drawer_button')).tap();
+        await expect(element(by.id('channels_list'))).toBeVisible();
+
+        // # Verify unread channel(s) display at top of channel list (with mentions first, if any), in alphabetical order, with title "Unreads"
+        await expect(element(by.text('UNREADS'))).toBeVisible();
+        await expect(element(by.id('channel_item_display_name')).atIndex(0)).toHaveText(newChannel.display_name);
```
Comment (saturninoabril) [inline detox/e2e/test/messaging/message_posting.e2e.js:40]: As suggested by Joseph, please add another channel so we can validate that the channel in unread section is really sorted.

**Review by saturninoabril — CHANGES_REQUESTED** [review]
Thanks @clarmso. Looking good. Please see comments.

File: detox/e2e/test/messaging/message_posting.e2e.js (hunk 2 above)
Comment (josephbaylon) [reply detox/e2e/test/messaging/message_posting.e2e.js:52]: Hi @clarmso I believe you added `testID='channels_list'` to the title of the list and not to the enclosing container of the list (`channels_list.js`). I added it to the container as shown below and both assertions passed for me locally,

Screen Shot 2020-09-28 at 11 31 05 AM

```
await expect(element(by.id('channels_list').withDescendant(by.id('channel_item_display_name')))).toExist();
await expect(element(by.text(newChannel.display_name).withAncestor(by.id('channels_list')))).toExist();
```

File: app/components/sidebars/main/channels_list/channels_list.js [hunk 4]
```diff
@@ -129,7 +129,10 @@ export default class ChannelsList extends PureComponent {
         const leftComponent = onShowTeams ? <SwitchTeamsButton onShowTeams={onShowTeams}/> : null;
 
         const title = (
-            <View style={[styles.searchContainer, padding(isLandscape)]}>
+            <View
+                style={[styles.searchContainer, padding(isLandscape)]}
+                testID='channels_list'
```
[the remark this answers is not in the document]
Comment (josephbaylon) [reply app/components/sidebars/main/channels_list/channels_list.js:134]: `testID='channels_list'` needs to be moved to enclosing container (see my other comment); currently inside list title

File: detox/e2e/test/messaging/message_posting.e2e.js (hunk 1 above)
Comment (clarmso) [reply detox/e2e/test/messaging/message_posting.e2e.js]: In addition, the "Unreads" should be at the top. I think this heading is displayed using `/app/components/sidebars/main/channels_list/filtered_list/filtered_list.js`. In particular, "Unreads" may be rendered via `renderSectionHeader`.  I have tried adding `testID` under `<View>` and `<Text>` but have not gotten the desired effect.

Verifying the "Unreads" exists is weaker than verifying it's the first heading. :(

File: detox/e2e/test/smoke_test/channels.e2e.js [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        // # Open channel drawer (with at least one unread channel)
+        await element(by.id('channel_drawer.button')).tap();
+
+        // # Verify unread channel(s) display at top of channel list (with mentions first, if any), in alphabetical order, with title "Unreads"
+        await expect(element(by.text('UNREADS'))).toBeVisible();
+        await expect(element(by.id('channel_item.display_name').withAncestor(by.id('channels_list'))).atIndex(0)).toHaveText(aChannel.display_name);
+        await expect(element(by.id('channel_item.display_name').withAncestor(by.id('channels_list'))).atIndex(1)).toHaveText(newChannel.display_name);
+        await expect(element(by.id('channel_item.display_name').withAncestor(by.id('channels_list'))).atIndex(2)).toHaveText(zChannel.display_name);
+
+        // # Tap an unread channel to view it
```
Comment (josephbaylon) [inline detox/e2e/test/smoke_test/channels.e2e.js:40]:
```
        // # Tap each unread channel to view all of them
        await element(by.text(aChannel.display_name)).tap();
        await element(by.id('channel_drawer.button')).tap();
        await element(by.text(newChannel.display_name)).tap();
        await element(by.id('channel_drawer.button')).tap();
        await element(by.text(zChannel.display_name)).tap();
```

File: detox/e2e/test/smoke_test/channels.e2e.js [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+        await element(by.text(newChannel.display_name)).tap();
+
+        // # Open channel drawer again
+        await element(by.id('channel_drawer.button')).tap();
+
+        // * Channel you just read is no longer listed in Unreads
```
Comment (josephbaylon) [inline detox/e2e/test/smoke_test/channels.e2e.js:46]: Verify that channels read are no longer in unreads

```
        // * Channels you just read are no longer listed in Unreads
        await expect(element(by.text('UNREADS'))).not.toBeVisible();
        await expect(element(by.text('PUBLIC CHANNELS'))).toBeVisible();
        await expect(element(by.id('channel_item.display_name').withAncestor(by.id('channels_list'))).atIndex(0)).toHaveText(aChannel.display_name);
        await expect(element(by.id('channel_item.display_name').withAncestor(by.id('channels_list'))).atIndex(1)).toHaveText(newChannel.display_name);
        await expect(element(by.id('channel_item.display_name').withAncestor(by.id('channels_list'))).atIndex(2)).toHaveText('Off-Topic');
        await expect(element(by.id('channel_item.display_name').withAncestor(by.id('channels_list'))).atIndex(3)).toHaveText('Town Square');
        await expect(element(by.id('channel_item.display_name').withAncestor(by.id('channels_list'))).atIndex(4)).toHaveText(zChannel.display_name);
```

**Review by josephbaylon — CHANGES_REQUESTED** [review]
@clarmso I would add additional actions and verifications as indicated in my suggestion to make sure unread channels no longer appear after being visited. See comments below. Thanks

**Review by josephbaylon — APPROVED** [review]
LGTM. Test passed locally. Thanks @clarmso !!

File: detox/e2e/test/smoke_test/channels.e2e.js [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+        let ch;
+
+        ch = await Channel.apiCreateChannel({type: 'O', prefix: 'a-channel', teamId: team.id});
+        aChannel = ch.channel;
+        await Channel.apiAddUserToChannel(user.id, aChannel.id);
+
+        ch = await Channel.apiCreateChannel({type: 'O', prefix: 'z-channel', teamId: team.id});
+        zChannel = ch.channel;
+        await Channel.apiAddUserToChannel(user.id, zChannel.id);
```
Comment (saturninoabril) [inline detox/e2e/test/smoke_test/channels.e2e.js:25]: nit: Looks weird to me. Got me thinking on whether we need to revisit on how the API calls are returning data. How about like below?
```suggestion
        ({channel: aChannel} = await Channel.apiCreateChannel({type: 'O', prefix: 'a-channel', teamId: team.id}));
        await Channel.apiAddUserToChannel(user.id, aChannel.id);

        ({channel: zChannel} = await Channel.apiCreateChannel({type: 'O', prefix: 'z-channel', teamId: team.id}));
        await Channel.apiAddUserToChannel(user.id, zChannel.id);
```

**Review by saturninoabril — COMMENTED** [review]
@clarmso Thanks for updating. Overall test flow looks good to me. Just last request and I'm all good.

**Review by saturninoabril — APPROVED** [review]
Thanks @clarmso, LGTM!

File: app/components/sidebars/main/channels_list/channels_list.js [hunk 8]
```diff
@@ -129,7 +129,9 @@ export default class ChannelsList extends PureComponent {
         const leftComponent = onShowTeams ? <SwitchTeamsButton onShowTeams={onShowTeams}/> : null;
 
         const title = (
-            <View style={[styles.searchContainer, padding(isLandscape)]}>
+            <View
+                style={[styles.searchContainer, padding(isLandscape)]}
```
Comment (enahum) [inline app/components/sidebars/main/channels_list/channels_list.js:133]: unnecessary change but non blocking

**Review by enahum — APPROVED** [review]
🎉
